extensions/EXT/EGL_EXT_present_opaque.txt

```diff
@@ -0,0 +1,109 @@
+Name
+
+    EXT_present_opaque
+
+Name Strings
+
+    EGL_EXT_present_opaque
+
+Contributors
+
+    Eric Engestrom
+
+Contacts
+
+    Eric Engestrom (eric 'at' engestrom.ch)
+
+IP Status
+
+    No known IP claims.
+
+Status
+
+    Draft
+
+Version
+
+    #1, August 8, 2021
+
+Number
+
+    EGL Extension #146
+
+Dependencies
+
+    Requires EGL 1.4 and EGL_KHR_image_base extension
+
+    This extension is written against the wording of the EGL 1.4
+    specification, and EGL_KHR_image_base version 6.
+
+Overview
+
+    This extension adds a new EGL surface attribute EGL_PRESENT_OPAQUE_EXT
+    to indicate that the surface should be presented as opaque,
+    disregarding any alpha channel if present.
+    If surface attribute EGL_PRESENT_OPAQUE_EXT is EGL_TRUE, then the
+    surface will be presented as opaque.
+
+New Types
+
+    None
+
+New Procedures and Functions
+
+    None
+
+New Tokens
+
+    New EGLSurface attribute name:
+
+        EGL_PRESENT_OPAQUE_EXT                  0x31DF
+
+
+Additions to Chapter 3 of the EGL 1.4 Specification (Rendering Surfaces)
+
+    Change the second paragraph in section 3.5 on p. 28 (describing
+    eglCreateWindowSurface):
+
+        "Attributes that can be specified in attrib list include EGL_RENDER_BUFFER,
+        EGL_PRESENT_OPAQUE_EXT, EGL_VG_COLORSPACE, and EGL_VG_ALPHA_FORMAT."
+
+    Add the following paragraph in section 3.5 on p. 28 before
+    "EGL_VG_COLORSPACE specifies the color space used by OpenVG"
+    (describing eglCreateWindowSurface attrib_list):
+
+        "EGL_PRESENT_OPAQUE_EXT specifies the presentation opacity mode
+        of the window surface. If its value is EGL_TRUE, then the
+        surface's alpha channel (if any) will be ignored and considered
+        fully opaque. If its value is EGL_FALSE, then the surface's
+        alpha channel is respected by the compositor. The default value
+        of EGL_PRESENT_OPAQUE_EXT is EGL_FALSE."
```

Comment:
I think what we would do is only advertise this on wayland-platform EGLDisplays, but is there any concern that when this extension is advertised, apps that set EGL_PRESENT_OPAQUE_EXT to EGL_FALSE (or leave it at the default) and rely on native winsys mechanisms to actually present transparent/translucent content will be violating the spec? Or is the spec intended to override the native mechanism in that case?

Reply:
I'm not sure what you mean here - can you please reword?

Reply:
I don't quite understand your question either, but I tried to word it as:
> `EGL_FALSE` => winsys behaviour is unchanged (ie. if it normally ignores the alpha channel it will keep doing that, and if it normally uses it it wil keep doing that

and
> `EGL_TRUE` => winsys pretends alpha is 1 (or whatever value "opaque" is)

I'm happy to change the text if you think that's not clear :)

Reply:
FALSE is the default. There is no conformant Wayland implementation that acts as if it were true.

Given the influence of visuals and composition, I don’t think this could be made to work at all correctly under X11. It might be relevant for EGLOutput to display directly I guess?

Reply:
Sorry if I wasn't clear. It is indeed the X11  case I had in mind. If we exposed this extension on X11, I don't know how we would implement it. If the app set EGL_PRESENT_OPAQUE_EXT to EGL_TRUE, we could hypothetically override SwapBuffers to write 1.0 to the alpha channel, and EGL doesn't have front-buffered rendering, so that would solve it, but it's unfortunate to need that invasive of a change for this simple extension.

Regardless, what I'm more worried about is an app selecting an X visual that has no alpha channel, then setting this to EGL_FALSE, or more realistically, leaving it at the default. My reading of the above sentence, "If its value is EGL_FALSE, then the surface's alpha channel is respected by the compositor," was that the compositor must use the alpha channel provided by the app. However, the compositor doesn't talk to the app directly, nor vice-versa, so it has to rely on the window's visual to determine whether there is an alpha channel, and that's going to say there isn't one, so it will behave as if EGL_PRESENT_OPAQUE_EXT were set to EGL_TRUE.

The plan is that we're just not going to advertise this for X11 apps, so perhaps the outcome is just academic. I was just curious whether others planned to advertise this on X11 or similar environments, and hence were interesting in a more relaxed wording corresponding to what you list as your intent above, that EGL_FALSE == status quo.

Reply:
> The plan is that we're just not going to advertise this for X11 apps, so perhaps the outcome is just academic. 

Yeah, I don’t think you can. On X11 this flag is effectively determined by the native visual of the config you select - which isn’t a concept we have on Wayland, hence why we need to provide something new to give apps the same behaviour they get by applying an alphaful config to a non-alphaful visual …

> we could hypothetically override SwapBuffers to write 1.0 to the alpha channel

… which is why I don’t think this could work, because in the presence of buffer_age you would’ve destroyed visible and meaningful data.

In both cases (alphaful config with either Wayland and this property, or X11 and a non-alpha visual) we get the same behaviour: client gets to write to and read back from the alpha channel of the default framebuffer, but the presentation engine treats it as if it were constant 1.0.

Reply:
How about this wording?
> If its value is EGL_FALSE, then the compositor doesn't change its behaviour, and considers the surface's alpha channel the same way as if the extension wasn't implemented.

And yeah, for Mesa I don't think we'll have any reason to implement this on our other platforms than Wayland; you'll probably want to do the same in the NVIDIA driver.

Reply:
That'll work, yeah.

Reply:
Wording WFM.